File: python-algorithm/ds_4_linkedlist/linkedlist_4_others/L725-m.cpp
//
// Created by LiHeng on 2020/7/19.
//

#include <iostream>
#include <vector>
using namespace std;

struct ListNode {
    int val ;
    ListNode* next ;
    ListNode(int x): val(x) , next(NULL) {}
};

class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* root, int k) {
        vector<ListNode*> res(k, NULL);
        int len = 0;
        ListNode* cnt = root;
        while(cnt){
            ++len;
            cnt = cnt->next;
        }
        int size = len/k;
        int mod = len%k;
        ListNode* start = root;
        ListNode* cur = root;
        //下面的设计比较巧妙，自己没想到
        for (int i = 0; cur != NULL && i < k; i++) {
            res[i] = cur;
            int curSize = size + (mod-- > 0 ? 1 : 0);
            for (int j = 0; j < curSize - 1; j++) {
                cur = cur->next;
            }
            ListNode* next = cur->next;
            cur->next = NULL;
            cur = next;
        }
        return res;
    }

};

File: python-algorithm/ds_4_linkedlist/linkedlist_4_others/L725-m.py
#--coding:utf-8--
'''
@Time   : 2020/8/14
@Author : Heng Li
@Email  : liheng_lulia@163.com
'''

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def splitListToParts(self, root: ListNode, k: int) -> List[ListNode]:
        res = [None for i in range(k)]
        if not root:
            return res
        tail = root
        lenth = 1
        while tail.next:
            tail = tail.next
            lenth += 1
        base = lenth//k
        addtion = lenth%k
        cur = root
        for i in range(k):
            if cur:
                res[i] = cur
                curSize = base + (1 if addtion >= 1 else 0)
                for j in range(curSize-1):
                    cur = cur.next
                new_start = cur.next
                cur.next = None # 断开
                cur = new_start
                addtion -= 1
        return res